### PR TITLE
[Example] Graph Random Neural Network

### PR DESCRIPTION
**Pull request by hengruizhang98**
## Description
This PR is the implementation of the paper: Graph Random Neural Networks for Semi-Supervised Learning on Graphs.
Current status:
1. Experiments on Cora, CiteSeer, Pubmed are conducted.
2. Experimental results are consistent with those reported in this paper.

TODO:
1. Try larger benchmarks.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred to in this PR

### CODE REVIEW

**Comment by mufeili** [comment]
Can you also add entries in examples/README.md for indexing? 

@VoVAllen Can we update the PR template to remind people about example indexing?

**Comment by hengruizhang98** [comment]
> Can you also add entries in examples/README.md for indexing?
> 
> @VoVAllen Can we update the PR template to remind people about example indexing?


Yes, I've added indexing for grand.

File: examples/pytorch/grand/main.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
+import dgl.function as fn
+def GRANDConv(graph, feats, order):
+
```
Comment (zhjwy9343) [inline examples/pytorch/grand/main.py:3]: As this is one of the key components of this paper, could you leave some comments about the function, such as what the input arguments are, what is it return, and what is the major algorithm used.

File: examples/pytorch/grand/main.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        for i in range(order):
+            graph.ndata['h'] = x
+            graph.update_all(fn.u_mul_e('h', 'weight', 'm'), fn.sum('m', 'h'))
+            x = graph.ndata.pop('h')
+            y.add_(x)
+
+    return x
+
+
+import argparse
```
Comment (zhjwy9343) [inline examples/pytorch/grand/main.py:24]: In Python coding style, we normally put all import clauses at the very beginning of a script.

File: examples/pytorch/grand/main.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+                    help='dropout rate of hidden layer')
+    parser.add_argument('--order', type=int, default=8, help='Propagation step')
+    parser.add_argument('--sample', type=int, default=4, help='Sampling times of dropnode')
+    parser.add_argument('--tem', type=float, default=0.5, help='Sharpening temperature')
+    parser.add_argument('--lam', type=float, default=1., help='Lamda')
+    parser.add_argument('--use_bn', action='store_true', default=False, help='Using Batch Normalization')
+
+    args = parser.parse_args()
+    if args.gpu != -1 and th.cuda.is_available():
+        args.device = 'cuda:%s'%args.gpu
```
Comment (zhjwy9343) [inline examples/pytorch/grand/main.py:69]: In Python 3, %s usage has been deprecated. Use string.format instead.

File: examples/pytorch/grand/main.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    else:
+        args.device = 'cpu'
+
+    return args
+
+if __name__ == '__main__':
+
```
Comment (zhjwy9343) [inline examples/pytorch/grand/main.py:76]: Please refer to the MVPExampleCode's main.py to add some comments of your main function.

File: examples/pytorch/grand/main.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+                  node_dropout, input_droprate, 
+                  hidden_droprate, batchnorm)
+
+    
+    model = model.to(device)
+    graph = graph.to(device)
+    
+
+    loss_fn = nn.NLLLoss()
+    opt = optim.Adam(model.parameters(), lr = args.lr, weight_decay = args.weight_decay)
```
Comment (zhjwy9343) [inline examples/pytorch/grand/main.py:130]: Suggest to either use all argument as args.*, or assign argument as in line 107 to 118.

File: examples/pytorch/grand/model.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
+import torch as th
+import torch.nn as nn
+import dgl.function as fn
+
+from layers import drop_node, MLP
+
+def GRANDConv(graph, feats, order):
```
Comment (zhjwy9343) [inline examples/pytorch/grand/model.py:7]: Any reason that you write two GRANDConv? I see one at the main.py file, and another one here.

File: examples/pytorch/grand/utils.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import torch as th
+import torch.nn as nn
+import torch.nn.functional as F
+
+def consis_loss(logps, temp, lam):
```
Comment (zhjwy9343) [inline examples/pytorch/grand/utils.py:5]: Is this function from the author? If so, pls add credit of the original authors.

File: examples/pytorch/grand/utils.py [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+    ps = [th.exp(p) for p in logps]
+    sum_p = 0.
+    for p in ps:
```
Comment (zhjwy9343) [inline examples/pytorch/grand/utils.py:8]: It seems this part does not fully leverage pytorch's vectorized operation. Please consider tune for loops into some pytorch tensor operations.

File: examples/pytorch/grand/model.py (hunk 6 above)
Comment (hengruizhang98) [reply examples/pytorch/grand/model.py:54]: Sorry, I just forgot to remove these lines.

File: examples/pytorch/grand/README.md [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+--sample           int     Sampling times of dropnode.            Default is 4.
+--tem              float   Sharpening temperaturer.               Default is 0.5.
+--lam              float   Lamda.                                 Default is 1.0.
+--use_bn           bool    Using batch normalization.             Default is False
+```
+
+## Examples
+
+The following commands learn a neural network and predict on the test set.
+Training a MixHop model on the default dataset.
```
[the remark this answers is not in the document]
Comment (hengruizhang98) [reply examples/pytorch/grand/README.md]: It should be GRAND

File: examples/pytorch/grand/main.py (hunk 1 above)
Comment (hengruizhang98) [reply examples/pytorch/grand/main.py]: I used this function for debugging but forgot to remove it.  now it's removed.

**Comment by VoVAllen** [comment]
@mufeili Template is at https://github.com/dmlc/dgl/tree/master/.github

**Comment by zhjwy9343** [comment]
After review of changes, I am ok with this PR.
